--- a/bitbake/lib/bb/runqueue.py
+++ b/bitbake/lib/bb/runqueue.py
@@ -1640,6 +1640,7 @@ class RunQueue:
                     logger.info("Tasks Summary: Attempted %d tasks of which %d didn't need to be rerun and %d failed.", self.rqexe.stats.completed + self.rqexe.stats.failed, self.rqexe.stats.skipped, self.rqexe.stats.failed)
                 else:
                     # Let's avoid the word "failed" if nothing actually did
+                    self.write_indexfile()
                     logger.info("Tasks Summary: Attempted %d tasks of which %d didn't need to be rerun and all succeeded.", self.rqexe.stats.completed, self.rqexe.stats.skipped)
 
         if self.state is runQueueFailed:
@@ -1848,6 +1849,15 @@ class RunQueue:
                 output = bb.siggen.compare_sigfiles(latestmatch, match, recursecb)
                 bb.plain("\nTask %s:%s couldn't be used from the cache because:\n  We need hash %s, most recent matching task was %s\n  " % (pn, taskname, h, prevh) + '\n  '.join(output))
 
+    def write_indexfile(self, indexfile_name='index.txt'):
+        directory = self.cfgData.getVar("SSTATE_DIR")
+        output_file = os.path.join(directory, indexfile_name)
+        with open(output_file, 'w') as f:
+            for root, dirs, files in os.walk(directory):
+                for file in files:
+                    file_path = os.path.join(root, file)
+                    relative_path = os.path.relpath(file_path, directory)
+                    f.write(f"{relative_path}\n")
 
 class RunQueueExecute:
 
