diff --git a/bitbake/lib/bb/runqueue.py b/bitbake/lib/bb/runqueue.py
index ffb2d28494..b33fb696c7 100644
--- a/bitbake/lib/bb/runqueue.py
+++ b/bitbake/lib/bb/runqueue.py
@@ -163,6 +163,12 @@ class RunQueueScheduler(object):
 
         self.rev_prio_map = None
         self.is_pressure_usable()
+        self.queue = open('./queue', 'a')
+        self.skip = open('./skip', 'a')
+
+    def __del__(self):
+        self.queue.close()
+        self.skip.close()
 
     def is_pressure_usable(self):
         """
@@ -244,12 +250,15 @@ class RunQueueScheduler(object):
         buildable.difference_update(self.rq.holdoff_tasks)
         buildable.intersection_update(self.rq.tasks_covered | self.rq.tasks_notcovered)
         if not buildable:
+            self.skip.write(str(time.time()) + ': ' + 'no buildable tasks\n')
             return None
 
-        # Bitbake requires that at least one task be active. Only check for pressure if
-        # this is the case, otherwise the pressure limitation could result in no tasks
-        # being active and no new tasks started thereby, at times, breaking the scheduler.
+        self.queue.write(str(time.time()) + '_buildable: ')
+        self.queue.writelines(list(str(self.buildable) + '\n'))
+        self.queue.write('\n')
         if self.rq.stats.active and self.exceeds_max_pressure():
+            self.skip.write(str(time.time()) + ': ')
+            self.skip.write('Pressure exceeded\n')
             return None
 
         # Filter out tasks that have a max number of threads that have been exceeded
@@ -269,10 +278,15 @@ class RunQueueScheduler(object):
             tid = buildable.pop()
             taskname = taskname_from_tid(tid)
             if taskname in skip_buildable and skip_buildable[taskname] >= int(self.skip_maxthread[taskname]):
+                self.skip.write(str(time.time()) + ': ')
+                self.skip.write(f'{taskname}:too much threads\n')
                 return None
             stamp = self.stamps[tid]
             if stamp not in self.rq.build_stamps.values():
                 return tid
+            else:
+                self.skip.write(str(time.time()) + ': ')
+                self.skip.write(f'{taskname}:task in buildstamps\n')
 
         if not self.rev_prio_map:
             self.rev_prio_map = {}
@@ -286,9 +300,13 @@ class RunQueueScheduler(object):
             if bestprio is None or bestprio > prio:
                 taskname = taskname_from_tid(tid)
                 if taskname in skip_buildable and skip_buildable[taskname] >= int(self.skip_maxthread[taskname]):
+                    self.skip.write(str(time.time()) + ': ')
+                    self.skip.write(f'{taskname}:too much threads\n')
                     continue
                 stamp = self.stamps[tid]
                 if stamp in self.rq.build_stamps.values():
+                    self.skip.write(str(time.time()) + ': ')
+                    self.skip.write(f'{taskname}:task in buildstamps\n')
                     continue
                 bestprio = prio
                 best = tid
@@ -301,6 +319,8 @@ class RunQueueScheduler(object):
         """
         if self.rq.can_start_task():
             return self.next_buildable_task()
+        else:
+            self.skip.write(str(time.time()) + ': too much threads\n')
 
     def newbuildable(self, task):
         self.buildable.add(task)
